cli/download_source.go

```diff
@@ -95,12 +105,17 @@ func downloadTerraformSourceIfNecessary(terraformSource *tfsource.TerraformSourc
 		return err
 	}
 
-	initFile := util.JoinPath(terraformSource.WorkingDir, moduleInitRequiredFile)
-	f, createErr := os.Create(initFile)
-	if createErr != nil {
-		return createErr
+	currentVersion := terraformSource.EncodeSourceVersion()
+	// if source versions are different, create file to run init
+	// https://github.com/gruntwork-io/terragrunt/issues/1921
+	if previousVersion != currentVersion {
```

Comment:
I'm not sure I follow the logic... Why is checking for `source`  equality the solution here, if the original problem was that if you didn't have `module` in your code, our `.modules` check was incorrect?

Reply:
Creation of `module` directory is just a workaround to skip initialization, original issue occurs because "local" dependencies are copied each time to `.terragrunt-cache`, and to know when to do initialization or not - I added logic to verify checksum of terraform version.

Reply:
The way I understand this logic now is:

Cases where it will not auto-init:

If there's no `modules` dir
and the terraformSources are the same
and there are no `module` blocks in the *.tf files

If there's a `modules` dir

Cases where it will auto-init:

If there's no `modules` dir
and the terraformSources are different

If there's no `modules` dir
and the terraformSources are different
but there are `module` blocks in the *.tf files

This change makes sense to me. In fact, I'm not fully understanding the current code as it is. Why does it always create the initFile?

Reply:
This PR has dependencies with another issue: https://github.com/gruntwork-io/terragrunt/issues/1901
`initFile` was introduced to automatically (re)run terragrunt init, with this PR it will compare source code versions before creation of the `initFile` - otherwise, for local dependencies, it will do init each time

Reply:
Ah I see.